ichnaea/service/locate.py

```diff
@@ -706,80 +762,85 @@ def __init__(self, db_sources, *args, **kwargs):
         self.all_providers = [
             cls(
                 db_sources[cls.db_source_field],
-                self.result_type,
                 api_key_log=self.api_key_log,
                 api_key_name=self.api_key_name,
                 api_name=self.api_name,
             ) for cls in self.provider_classes]
 
     def search_location(self, data):
-        result = self.result_type(None, query_data=False)
-        all_results = defaultdict(deque)
-
-        for location_provider in self.all_providers:
-            provider_result = location_provider.locate(data)
-            all_results[location_provider.log_group].appendleft(
-                provider_result)
-
-            if provider_result.more_accurate(result):
+        best_location = None
+        best_location_provider = None
+        all_locations = defaultdict(deque)
+
+        for provider in self.all_providers:
+            provider_location = provider.locate(data)
+            all_locations[
+                provider.log_group].appendleft(
+                (provider, provider_location))
+
+            if best_location is None or provider_location.more_accurate(
+                    best_location):
                 # If this location is more accurate than our previous one,
                 # we'll use it.
-                result = provider_result
+                best_location = provider_location
+                best_location_provider = provider
 
-            if result.accurate_enough():
-                # Stop the loop, if we have a good quality result.
+            if best_location.accurate_enough():
```

Comment:
Are we missing a test here where the first provider doesn't return a result at all? Shouldn't best_location still be None then here, so `best_location.accurate_enough()` would fail? I kinda liked the empty result idea, which avoided this sort of "is None" checking. And spelling that out as a concrete class, like you did in between looked pretty good to me.